### PR TITLE
Add contrail color customisation

### PR DESCRIPTION
**Pull request by PunkPun**
I add `EndColor` and Transparency functionality from #20037 as well as other minor tweaks not relating to width

### CODE REVIEW

**Review by AspectInteractive — DISMISSED** [review]
Having reviewed the original contrail PR everything here looks to be the same and is fine.

**Comment by dnqbob** [comment]
If you want to go further on color customisation, you can also:

1. Add `BlendMode` to Contrail. You need also add blendMode to `DrawLine` in `ContrailRendrable`. BlendMode default to `Alpha`(this will be a easier one, I think you can try it in this PR)

2. Add a new `DrawLine` that support color change like contrail color radius-changing, but this will be hard especially on the user interface naming(like `EnableRadiusChanging`, `RadiusStartColor`, `RadiusEndColor`). 

A new `DrawLine` also benefits the laser&beam rendarable later.

Contrail color radius-changing showcase (also its blendmode is Additive):
ta2

**Comment by dnqbob** [comment]
If you want to go further on color customisation, you can also:

1. Add blendeMode to `DrawLine` in `ContrailRendrable`, defualt to `Alpha`(this will be a easier one, I think you can try it in this PR)

2. Add a new `DrawLine` that support color change like contrail color radius-changing, but this will be hard especially on the user interface naming(like `EnableRadiusChanging`, `RadiusStartColor`, `RadiusEndColor`). 

A new `DrawLine` also benefits the laser&beam rendarable later.

Contrail color radius-changing showcase:
ta2

**Comment by PunkPun** [comment]
I'm not too familiar with `BlendMode` or any of the graphics code so I won't be doing that. If you want to you can create a PR for it, possibly having this PR as a dependency

File: OpenRA.Mods.Common/Projectiles/Bullet.cs [hunk 1]
```diff
@@ -101,13 +101,36 @@ public class BulletInfo : IProjectileInfo
...
+		[Desc("Length of the contrail (in ticks).")]
 		public readonly int ContrailLength = 0;
-		public readonly int ContrailZOffset = 2047;
-		public readonly Color ContrailColor = Color.White;
-		public readonly bool ContrailUsePlayerColor = false;
+
+		[Desc("Delay of the contrail (in ticks).")]
 		public readonly int ContrailDelay = 1;
+
+		[Desc("Offset for contrail's Z sorting.")]
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Projectiles/Bullet.cs]:
```suggestion
		[Desc("Equivalent to sequence ZOffset. Controls Z sorting.")]
```

File: OpenRA.Mods.Common/Projectiles/Bullet.cs [hunk 2]
```diff
@@ -101,13 +101,36 @@ public class BulletInfo : IProjectileInfo
...
+		public readonly int ContrailZOffset = 2047;
+
+		[Desc("Width of the contrail.")]
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Projectiles/Bullet.cs]:
```suggestion
		[Desc("Thickness of the emitted line.")]
```

Try to not just repeat variable name when describing it.

File: OpenRA.Mods.Common/Projectiles/Bullet.cs [hunk 3]
```diff
@@ -101,13 +101,36 @@ public class BulletInfo : IProjectileInfo
 		[Desc("Altitude above terrain below which to explode. Zero effectively deactivates airburst.")]
 		public readonly WDist AirburstAltitude = WDist.Zero;
 
+		[Desc("Length of the contrail (in ticks).")]
 		public readonly int ContrailLength = 0;
-		public readonly int ContrailZOffset = 2047;
-		public readonly Color ContrailColor = Color.White;
-		public readonly bool ContrailUsePlayerColor = false;
+
+		[Desc("Delay of the contrail (in ticks).")]
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Projectiles/Bullet.cs]:
```suggestion
		[Desc("Time (in ticks) after which the line should appear. Controls the distance to the actor.")]
```

File: OpenRA.Mods.Common/Projectiles/Bullet.cs [hunk 4]
```diff
@@ -101,13 +101,36 @@ public class BulletInfo : IProjectileInfo
 		[Desc("Altitude above terrain below which to explode. Zero effectively deactivates airburst.")]
 		public readonly WDist AirburstAltitude = WDist.Zero;
 
+		[Desc("Length of the contrail (in ticks).")]
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Projectiles/Bullet.cs]:
```suggestion
		[Desc("When set, display a line behind the actor that fades out in the end. Length is measured in ticks after appearing.")]
```

File: OpenRA.Mods.Common/Projectiles/Bullet.cs [hunk 5]
```diff
@@ -101,13 +101,36 @@ public class BulletInfo : IProjectileInfo
...
 		public readonly WDist ContrailWidth = new WDist(64);
 
+		[Desc("RGB color at the contrail start.")]
+		public readonly Color ContrailStartColor = Color.White;
+
+		[Desc("Use player remap color instead of a custom color at the contrail the start.")]
+		public readonly bool ContrailStartColorUsePlayerColor = false;
+
+		[Desc("The alpha value [from 0 to 255] of color at the contrail the start.")]
+		public readonly int ContrailStartColorAlpha = 255;
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Projectiles/Bullet.cs:123]: You don't need this.

> Color (RRGGBB[AA] notation)

This can be defined in `ContrailStartColor`.

Comment (dnqbob) [reply OpenRA.Mods.Common/Projectiles/Bullet.cs:123]: I am afraid we still need this, considering we can also set up player color and we want to add an alpha control to it

Comment (dnqbob) [reply OpenRA.Mods.Common/Projectiles/Bullet.cs:123]: For example, we fade with player color but alpha=0

Comment (Mailaender) [reply OpenRA.Mods.Common/Projectiles/Bullet.cs:123]: I see.

File: OpenRA.Mods.Common/Projectiles/Bullet.cs (hunk 4 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Projectiles/Bullet.cs]: Changed it to `When set, display a line behind the actor. Length is measured in ticks after appearing.`

**Review by Mailaender — DISMISSED** [review]

**Review by dragunoff — CHANGES_REQUESTED** [review]
Setting a start color in this PR does produce the same result as setting the color before this PR. By setting only start color we get a contrail that transitions from color to white and then to transparency. Before this PR the contrail transitioned from color to transparency.

This PR:
```miniyaml
ContrailStartColor: FF0000
```
Screenshot from 2022-09-23 17-41-35

Before:
```miniyaml
ContrailColor: FF0000
```
Screenshot from 2022-09-23 17-38-47

To achive the same effect with this PR both start and end color need to be the same:
```miniyaml
ContrailStartColor: FF0000
ContrailEndColor: FF0000
```
image

So the update rule here needs to also set the end color if we want to ensure the same effect. Or the logic should be changed so that specifying only one of the colors also sets the other.

**Comment by PunkPun** [comment]
>  By setting only start color we get a contrail that transitions from color to white and then to transparency.

That's simply not how this works. It transitions to a new color and new transparency at the same time, it may just not look like it when the transparency is set to 0.

> So the update rule here needs to also set the end color if we want to ensure the same effect. Or the logic should be changed so that specifying only one of the colors also sets the other.

I took a more modder friendly approach. If end color is not set, it's now equal to start color. We keep the same logic and reduce the amount of redundant yaml definitions

**Comment by dragunoff** [comment]
> 
> That's simply not how this works. It transitions to a new color and new transparency at the same time, it may just not look like it when the transparency is set to 0.
> 

Of course, I just described it so poorly :zany_face: :wink: 

> 
> I took a more modder friendly approach. If end color is not set, it's now equal to start color. We keep the same logic and reduce the amount of redundant yaml definitions
>

Nice, I think that's the better approach :+1:

File: OpenRA.Mods.Common/Projectiles/Bullet.cs (hunk 4 above)
Comment (dragunoff) [reply OpenRA.Mods.Common/Projectiles/Bullet.cs]: Length here means "duration" and not "distance" even though that's probably how people would think about this. I guess that description is OK :shrug:

**Review by dragunoff — DISMISSED** [review]

**Comment by PunkPun** [comment]
removed the testcase

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/7fe57b677f1f00fccd33fdda7f0ed5f364483522...0c69491198d0fc0f392761f9a4902352cab6f57f)

**Comment by dnqbob** [comment]
I will then add the blendmode for contrail

**Comment by Mailaender** [comment]
Note: the update rule does not cover weapons.

**Comment by PunkPun** [comment]
It should?

File: OpenRA.Mods.Common/UpdateRules/Rules/20210321/RenameContrailProperties.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+			foreach (var traitNode in actorNode.ChildrenMatching("Contrail"))
+				traitNode.RenameChildrenMatching("UsePlayerColor", "StartColorUsePlayerColor");
+
+			yield break;
+		}
+
+		public override IEnumerable<string> UpdateWeaponNode(ModData modData, MiniYamlNode weaponNode)
+		{
+			foreach (var traitNode in weaponNode.ChildrenMatching("Missile"))
```
Comment (Mailaender) [inline OpenRA.Mods.Common/UpdateRules/Rules/20210321/RenameContrailProperties.cs:35]: This should be `weaponNode`.

**Pull request merged**